### DISCUSSION TITLE
Listening to tokio_uring::net::TcpListener

### DISCUSSION THREAD

**Discussion by milen-denev** [tokio-rs/axum#2989] [Q&A]
I was wondering if the new #2941 pull request will allow to use the io_uring TcpListener and if even will have any performance benefits, thanks to the team!

**Comment by mladedav** [comment]
Yes, I believe it should be possible. I'll try and check if I get some time.

We'll see if it will have impact on performance, but hopefully it could be fatser. I didn't follow io_uring for some time but I know there were some security issues and I'd be more concerned with stability at this point though.

**Reply by mladedav** [reply]
Trying this out, it might not be possible yet, because we make use of `AsyncRead` and `AsyncWrite`, but `tokio_uring::net::TcpStream` does not seem to implement these. Their [design doc](https://github.com/tokio-rs/tokio-uring/blob/7761222aa7f4bd48c559ca82e9535d47aac96d53/DESIGN.md?plain=1#L495) suggests that it will implement them though.

**Reply by milen-denev** [reply]
Thank you very much and I would wait with excitement about that, should I mark as answered you will post any update in the future?

**Reply by mladedav** [reply]
I'll update here in the future.

**Reply by mladedav** [reply]
I don't think it will be possible to use `tokio_uring` at this time. It's built for single-threaded runtime usage and because of reasons described in the design doc mentioned earlier, many of the primites are `!Send`.

`axum` spawns a task per connection and since these `tokio` uses work-stealing runtime, all of these tasks need to be `Send`. Because they internally use the custom `Listener` and the `Io` type it produces, these also need to be `Send`.

**Reply by mladedav** [reply]
[hyper's issue](https://github.com/hyperium/hyper/issues/2140) on `io_uring` might also be of interest to you.

**Reply by milen-denev** [reply]
Ok thank you for the more details on the topic, I wasn't aware of these specific details. I will look myself more at that.
